src/storage/src/storage/bidi/worker.rs

```diff
@@ -0,0 +1,733 @@
+                        break;
+                    };
+                    self.insert_range(tx.clone(), range).await;
+                },
+            }
+        }
+        // No need to continue reading. The `requests` queue closes
+        // only when the ObjectDescriptor is gone and when all the
+        // associated ReadResponseReaders are gone.
+        Ok(())
+    }
+
+    async fn handle_response(
+        &mut self,
+        message: tonic::Result<Option<BidiReadObjectResponse>>,
+    ) -> Option<LoopResult<Option<Connection<C::Stream>>>> {
+        let response = match message.transpose()? {
+            Ok(r) => r,
+            Err(status) => return self.reconnect(status).await,
+        };
+
+        if let Err(e) = self.handle_ranges(response.object_data_ranges).await {
+            // An error in the response. These are not recoverable.
+            let error = Arc::new(e);
+            self.close_readers(error.clone()).await;
+            return Some(Err(error));
+        }
+        Some(Ok(None))
+    }
+
+    async fn handle_ranges(&self, data: Vec<ObjectRangeData>) -> crate::Result<()> {
+        let ranges = self.ranges.clone();
+        let pending = data
+            .into_iter()
+            .map(|r| Self::handle_range_data(ranges.clone(), r))
+            .collect::<Vec<_>>();
+        let _ = futures::future::join_all(pending)
+            .await
+            .into_iter()
+            .collect::<ReadResult<Vec<_>>>()
+            .map_err(crate::Error::io)?; // TODO: think about the error type
+        Ok(())
+    }
+
+    async fn reconnect(
+        &mut self,
+        status: tonic::Status,
+    ) -> Option<LoopResult<Option<Connection<C::Stream>>>> {
+        let ranges: Vec<_> = self
+            .ranges
+            .lock()
+            .await
+            .iter()
+            .map(|(id, r)| r.as_proto(*id))
+            .collect();
+        let (response, connection) = match self.connector.reconnect(status, ranges).await {
+            Err(e) => {
+                let error = Arc::new(e);
+                self.close_readers(error.clone()).await;
+                return Some(Err(error));
+            }
+            Ok((m, connection)) => (m, connection),
+        };
+        if let Err(e) = self.handle_ranges(response.object_data_ranges).await {
+            // An error in the response. These are not recoverable.
+            // TODO: refactor to handle_ranges().
+            let error = Arc::new(e);
+            self.close_readers(error.clone()).await;
+            return Some(Err(error));
+        }
+        Some(Ok(Some(connection)))
+    }
+
+    async fn close_readers(&mut self, error: Arc<crate::Error>) {
+        let mut guard = self.ranges.lock().await;
+        let closing: Vec<_> = guard
+            .iter_mut()
+            .map(|(_, pending)| pending.interrupted(error.clone()))
+            .collect();
+        let _ = futures::future::join_all(closing).await;
```

Comment:
no action required: Since ordering doesn't matter and if this could be a large number of futures, [FuturesUnordered](https://docs.rs/futures/0.3.31/futures/prelude/stream/struct.FuturesUnordered.html) may be more efficient